### DISCUSSION TITLE
Custom Cache Keys on Vercel CDN

### DISCUSSION THREAD

**Discussion by nickjackson** [vercel/community#86] [Idea]
We are building an e-commerce store using next.js for a client of ours and would like to cache the product page on Vercel's CDN but have a little stumbling block.

We use `getServerSideProps` on the product page to fetch the product details and were hoping to make use of `Cache-Control` headers to cache the page on Vercel's CDN. However, in this particular case, the e-commerce store allows their customers to negotiate the product price and they would like that discounted price to show on the product page instead (when they are logged in). As far as I'm aware this means we cannot cache this page on Vercel's edge because of the diverging nature of the SSR.

It would be great if we had some ability to customise the Cache Key on Vercel's CDN, to incorporate a cookie value (if present) so that we can have some advanced control over which version of a page is served from the cache.

**Comment by voronianski** [comment]
+1

**Comment by leerob** [comment]
We are exploring this, thank you for suggesting!

**Reply by benkingcode** [reply]
Any news on this? We simply can't use Vercel if we can't customise the cache key... any URL that has marketing query params on the end will result in 100% cache misses.

**Comment by leerob** [comment]
Cache tags now exists for Next.js on Vercel (currently just that framework, but would like to expand) through the Vercel Data Cache.

https://vercel.com/docs/infrastructure/data-cache#tag-based-revalidation

**Reply by bitttttten** [reply]
Hey @leerob ! Thanks for the reply here. I'm curious about this behaviour too, on our current website we have the auth cookie in the cache key so the _pages_ can be cached on the CDN, not just the `fetch` calls that run inside the function. So we have our homepage HTML cached for logged out users for each deployment.

Now we are on Vercel and looking to replicate this behaviour. Our current implementation (shared later) is marked as dynamic and we cannot cache it on the Edge Network Cache since we read from `cookies` inside the Server Component `page.tsx`, Vercel always sets it as MISS even though if you do not have the auth cookie it's totally cacheable.

Imagine this page below. `loggedin` and `loggedout` are slots. It's safe to cache the below page if the `auth` cookie does not exist. It's also safe to cache the page based on the `auth` cookie's encrypted value to.

```tsx
import "./globals.css";
import type { Metadata } from "next";
import { Inter } from "next/font/google";
import { cookies } from "next/headers";

const inter = Inter({ subsets: ["latin"] });

export default function RootLayout({
  children,
  loggedin,
  loggedout,
}: {
  children: React.ReactNode;
  loggedin: React.ReactNode;
  loggedout: React.ReactNode;
}): JSX.Element {
  const requestCookies = cookies();
  return (
    <html lang="en">
      <body className={inter.className}>
        {requestCookies.has("auth") ? loggedin : loggedout}
      </body>
    </html>
  );
}
```

Maybe this is possible in a Vercel way, or on Vercel. From my learnings I haven't come across the ability to do this yet.

Appreciate all the work and communication you are also giving Lee. Thanks 🙏

**Reply by leerob** [reply]
It might be better to redirect or rewrite at the CDN level (using the configuration options) based on the cookie? You would still verify the authentication / authorization state at the component level but this would allow you to serve up a static shell for a fast initial response, while still preventing the flash of content based on login state.

**Reply by bitttttten** [reply]
Appreciate the reply @leerob , sorry for the slow poke I was away on holiday!

I follow your logic although I am not sure where the variants or variations of the static shells would live 😅 Do you mind expanding on that? Maybe I am missing something in my knowledge gap here although let's say during `npm run build` how does Next.JS know to render two different variants of my above RootLayout? How do I tell it about `requestCookies.has("auth")`?

And if I am using Vercel's CDN, how would I configure that cache using the configuration options? Perhaps I'm also missing something in the Vercel docs although I can't find much about that!

I am imagining the scenarios:

1. User without an auth cookie, they get a cached response from the CDN (if it exists) otherwise we go to the origin Next.JS server and run the above.
2. User with an auth cookie, they don't hit the cache, and we can do the dynamic fetch here and do not cache.
